R/make-seurat.R

```diff
@@ -0,0 +1,135 @@
+      requireNamespace("Seurat", quietly = TRUE),
+    "Package `SeuratObject` must be installed for SCE to Seurat conversion." =
+      requireNamespace("SeuratObject", quietly = TRUE),
+    "`sce` must be a SingleCellExperiment object." = is(sce, "SingleCellExperiment"),
+    "`sce` must contain both `gene_ids` and `gene_symbol` columns in the row data if it is being used as a reference." =
+      reference != "sce" || all(c("gene_ids", "gene_symbol") %in% names(rowData(sce)))
+  )
+
+  if (use_symbols) {
+    sce <- suppressMessages(sce_to_symbols(
+      sce,
+      reference = reference,
+      unique = (dedup_method == "unique"),
+      seurat_compatible = TRUE
+    ))
+  }
+
+  if (dedup_method == "sum") {
+    sce <- sum_duplicate_genes(sce)
+  }
+
+  mainExpName(sce) <- "RNA"
+
+  # convert reducedDims to Seurat standard naming
+  reducedDimNames(sce) <- tolower(reducedDimNames(sce))
+  reducedDims(sce) <- reducedDims(sce) |>
+    purrr::map(\(mat) {
+      # insert underscore between letters and numbers in axis labels
+      rd_names <- gsub("([A-Za-z]+)([0-9]+)", "\\1_\\2", colnames(mat)) |>
+        tolower()
+      colnames(mat) <- rd_names
+      if (!is.null(attr(mat, "rotation"))) {
+        colnames(attr(mat, "rotation")) <- rd_names
+      }
+      return(mat)
+    })
+
+  # pull out any altExps to handle manually
+  alt_exps <- altExps(sce)
+  altExps(sce) <- NULL
+
+  # Let Seurat do initial conversion to capture most things
+  sobj <- Seurat::as.Seurat(sce)
+
+  # update identity values, using cluster if present
+  sce_meta <- metadata(sce)
+  sobj$orig.ident <- sce_meta$library_id
+  if (is.null(sce$cluster)) {
+    Seurat::Idents(sobj) <- sce_meta$library_id
+  } else {
+    Seurat::Idents(sobj) <- sce$cluster
+  }
+
+  # add variable features if present
+  if (!is.null(sce_meta$highly_variable_genes)) {
+    sobj[["RNA"]]@var.features <- sce_meta$highly_variable_genes
+    sce_meta$highly_variable_genes <- NULL # remove from metadata to avoid redundancy
+  }
+
+  # convert and set functions depending on assay version requested
+  if (seurat_assay_version == "v5") {
+    create_seurat_assay <- SeuratObject::CreateAssay5Object
+    suppressWarnings(sobj[["RNA"]] <- as(sobj[["RNA"]], "Assay5"))
+  } else {
+    create_seurat_assay <- SeuratObject::CreateAssayObject
+    suppressWarnings(sobj[["RNA"]] <- as(sobj[["RNA"]], "Assay"))
+  }
+
+  # add spliced data if present
+  if ("spliced" %in% assayNames(sce)) {
+    sobj[["spliced"]] <- create_seurat_assay(counts = assay(sce, "spliced"))
+  }
+
+  # add altExps as needed.
+  for (alt_exp_name in names(alt_exps)) {
+    alt_exp <- alt_exps[[alt_exp_name]]
+    stopifnot(
+      "All altExps must contain a `counts` assay." = "counts" %in% assayNames(alt_exp)
+    )
+    sobj[[alt_exp_name]] <- create_seurat_assay(counts = counts(alt_exp))
+    if ("logcounts" %in% assayNames(alt_exp)) {
+      sobj[[alt_exp_name]]$data <- logcounts(alt_exp)
```

Comment:
Worth noting that these lines may both give the Seurat warning
```
Warning: Feature names cannot have underscores ('_'), replacing with dashes ('-')
```

I got this when running a multiplexed library through, so this was coming from the `MULTI_X` names. We might want to wrangle the altexps a bit more first if we want to avoid this Seurat warning. I think would be good since it's not very transparent to users that _this_ is where the warning comes from, and we might add our own warning (message?) instead to be clearer what's changing.

Reply:
Added conversion for altExps with a warning.